doc/TAM/SAI-Proposal-TAM-stream-telemetry.md

```diff
@@ -0,0 +1,906 @@
+sai_attr_list[1].value.objlist.count = 1;
+sai_attr_list[1].value.objlist.list[0] = sai_tam_collector_obj;
+
+sai_attr_list[2].id = SAI_TAM_TELEMETRY_ATTR_REPORTING_TYPE;
+sai_attr_list[2].value.s32 = SAI_TAM_REPORTING_TYPE_COUNT_BASED
+
+sai_attr_list[3].id = SAI_TAM_TELEMETRY_ATTR_REPORTING_BULK_COUNT;
+sai_attr_list[3].value.u32 = bulk_count;
+
+attr_count = 4;
+
+sai_create_tam_telemetry_fn(&sai_tam_telemetry_obj, switch_id, attr_count, sai_attr_list);
+
+```
+
+#### Creating TAM counter subscription objects
+
+Based on the STREAM_TELEMETRY_GROUP on Config DB, to create corresponding counter subscription objects.
+
+``` c++
+
+/**
+ * @brief Counter Subscription attributes
+ */
+typedef enum _sai_tam_counter_subscription_attr_t
+{
+
+// ...
+
+    /**
+     * @brief Telemetry label
+     *
+     * Label to identify this counter in telemetry reports.
+     * If the report type is IPFIX, this label will be used as the element ID in the IPFIX template.
+     *
+     * @type sai_uint64_t
+     * @flags CREATE_ONLY
+     * @default 0
+     */
+    SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_LABEL,
+
+    /**
+     * @brief Setting of read-clear or read-only for statistics read.
+     *
+     * @type sai_stats_mode_t
+     * @flags CREATE_ONLY
+     * @default SAI_STATS_MODE_READ
+     */
+    SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_STATS_MODE,
+
+// ...
+
+} sai_tam_counter_subscription_attr_t;
+
+
+// Create counter subscription list
+
+sai_attr_list[0].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_TEL_TYPE;
+sai_attr_list[0].value.oid = sai_tam_tel_type_obj;
+
+sai_attr_list[1].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_OBJECT_ID;
+sai_attr_list[1].value.oid = port_obj;
+
+sai_attr_list[2].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_STAT_ID;
+sai_attr_list[2].value.oid = SAI_PORT_STAT_IF_IN_OCTETS;
+
+sai_attr_list[3].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_LABEL;
+sai_attr_list[3].value.oid = index; // Element ID of the object in the IPFIX template
+
+sai_attr_list[4].id =SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_STATS_MODE;
+sai_attr_list[4].value.s32 = SAI_STATS_MODE_READ;
+// sai_attr_list[4].value.s32 = SAI_STATS_MODE_READ_AND_CLEAR; // It could be read and clear for queue watermark
+
+attr_count = 5;
+
+create_tam_counter_subscription(&sai_tam_counter_subscription_obj, switch_id, attr_count, sai_attr_lis);
+// If this stats of object cannot support this poll frequency, this API should return SAI_STATUS_NOT_SUPPORTED.
+```
+
+#### Creating TAM object
+
+``` c++
+
+sai_attr_list[0].id = SAI_TAM_ATTR_TELEMETRY_OBJECTS_LIST;
+sai_attr_list[0].value.objlist.count = 1;
+sai_attr_list[0].value.objlist.list[0] = sai_tam_telemetry_obj;
+
+sai_attr_list[1].id = SAI_TAM_ATTR_TAM_BIND_POINT_TYPE_LIST;
+sai_attr_list[1].value.objlist.count = 2;
+sai_attr_list[1].value.objlist.list[0] = SAI_TAM_BIND_POINT_TYPE_PORT;
+sai_attr_list[1].value.objlist.list[0] = SAI_TAM_BIND_POINT_TYPE_QUEUE;
+
+attr_count = 2;
+sai_create_tam_fn(&sai_tam_obj, switch_id, attr_count, sai_attr_list);
+

```

Comment:
The sequence is incomplete. We need to bind sai\_tam\_obj to port/queue/switch etc. Pls refer SAI\_PORT\_ATTR\_TAM\_OBJECT.

Reply:
Yes, it's not a complete list. This list is fully depended on which counters we subscribed. 

That's why I said: 
https://github.com/opencomputeproject/SAI/pull/2089/files#diff-0e0ad14d84f3332634521f6b2d53007b4f7ccec6c53e003139ef77da1a4d9d50R373

BTW, why do you mention `SAI_PORT_ATTR_TAM_OBJECT`? I feel that's another communication mode for TAM. In our approach, we don't need to set this attribute.